### DISCUSSION TITLE
Empty code block included in Jupyter notebook when echo: false

### DISCUSSION THREAD

**Discussion by jreades** [quarto-dev/quarto-cli#10326] [Q&A] [answered]
### Description

Hi -- 

I don't know if this is something I'm doing wrong, an expected behaviour, or a bug, but I'm kind of surprised to see an empty code block output from the second call to `df.A.plot.hist()`. I would *expect* it to just output the image since `echo: false` and the code is unrunnable anyway since it was suppressed and replaced with the empty block.

````qmd
---
title: "Notebook Test"
subtitle: "Empty code blocks output"
jupyter: python3
---

This block works exactly as expected:

```{python}
import numpy as np
import pandas as pd

df = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))

df.A.plot.hist()
```

From this next block I would expect only to see the image, not the empty code block:

```{python}
#| echo: false
df.A.plot.hist()
```
````

Quarto version is 1.5.43 and running on OSX 14.5.

**Comment by mcanouil** [comment]
What do you want to do exactly?
`echo` is for source code.
It does not control outputs from your code.

See the documentation: https://quarto.org/docs/computations/execution-options.html

I suggest to use `keep-md: true` to investigate what exactly is the the thing you see.

I don't see anything weird.
<img width="594" alt="image" src="https://github.com/user-attachments/assets/cf56b189-ef9c-4376-8e70-8deade17eefa">

**Reply by mcanouil** [reply]
1.5.43 is a pre-release, I suggest to at least upgrade to the 1.5 stable release.

**Reply by jreades** [reply]
I have updated to 1.5.55 and still get the output I was perplexed by. I am not seeing the same thing that you are (see screenshot). Notice the code block `[2]` which doesn't appear in your image. 

![Screenshot 2024-07-19 at 09 42 14](https://github.com/user-attachments/assets/1182a431-306d-451e-9f70-8f99fb9e3502)

Will try a 1.6 pre-release quickly to see if that changes anything.

**Comment by jreades** [comment]
Upgrading Quarto hasn't changed the output so it must be something else. Is there a parameter I need to add to the render command perhaps?

`quarto render test.qmd --to ipynb`

Thank you.

**Reply by jreades** [reply]
The relevant markdown (after setting `keep-md` to true) section is:

````qmd
From this next block I would expect only to see the image, not the empty code block:

::: {#590ef959 .cell execution_count=2}

::: {.cell-output .cell-output-display}
![](test2_files/figure-ipynb/cell-3-output-1.png){}
:::
:::
````

I can't see where it would get the empty code block from which makes me wonder if it's JupyterLab that's the issue? I'm on Jupyter Lab 4.2.3. And Quarto 1.6.2 but going to roll that back to 1.5.55 now.

**Reply by mcanouil** [reply]
You are rendering to Jupyter Notebook? Next time, please include the format (or eventually the command you used) in your example from the start.
What do you want to achieve exactly? Adding a bit of context might help figure out the proper solution.

Note that I don't think it's possible to have the output of a code cell not attached to a code cell in Jupyter. (Look at the json produced in Jupyter for the relevant part of the document)

@cscheid might know better about this.

**Reply by jreades** [reply]
My apologies -- I didn't realise until this morning that I hadn't included the rendering command in my original question. 

I'm basically finishing up work on the [Q&A-type functionality](https://github.com/quarto-dev/quarto-cli/discussions/9739) for use in computer lab practicals where the questions are in a Jupyter notebook and the answers are in a PDF that I'll release later. These are done using a modified tabset:  

- Rendering to ipynb takes the first panel
- Rendering to pdf takes the second panel
- Rendering to html takes both panels

So I have that functionality working (see [github.com/jreades/qna](https://github.com/jreades/qna) and it can be installed via `quarto add ...` and was working at this point just to tidy things up. I thought 'Oh, it would be good to be able to say "This is what your output should look like" as part of the jupyter notebook' when I discovered that I always got the empty code block even though the image was created. Your comment that it might not be possible to have free-floating code outputs without a code block would make sense of what I'm seeing here.

**Reply by mcanouil** [reply, marked as answer]
Then you probably need to state that what your code cell emits is markdown and not the results of a Jupyter cell. `#| output: asis` is probably what you need (or a more complex thing using IPython).

**Reply by jreades** [reply]
FFS, I hadn't realised `asis` could be used that way. I had only used it for help with creating 'dynamic' Markdown. 

All sorted:

![Screenshot 2024-07-19 at 10 31 34](https://github.com/user-attachments/assets/4f7b9da5-6386-4bae-b0ee-3df1067cb6e0)

And for completeness (in case anyone else ever comes searching):

````qmd
---
title: "Notebook Test"
subtitle: "Suppressing code blocks that output images when rendering to Jupyter/iPython"
jupyter: python3
---

This block works exactly as expected:

```{python}
import numpy as np
import pandas as pd

df = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))

df.A.plot.hist()
```

And so do this one which shows what your 'answer' should look like:

```{python}
#| echo: false
#| output: asis
df.A.plot.hist()
```

````

**Reply by mcanouil** [reply]
FYI, the `output: asis` "trick" might not work for more complex output.
For those cases, additional handling will be required using IPython to emit proper Markdown.
